### PR TITLE
Improve performance of getSubstitutor

### PR DESCRIPTION
**Pull request by hit-lacus**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Branch to commit
- [ ] Branch **kylin3** for v2.x to v3.x
- [ ] Branch **kylin4** for v4.x
- [x] Branch **kylin5** for v5.x

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [ ] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin.apache.org or dev@kylin.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hit-lacus** [comment]
主要改动的是：
1. 只有 KylinExternalConfigLoader 实现 ICachedExternalConfigLoader，NacosExternalConfigLoader 还是原来的接口 
2. 修改 size 的 实现

File: src/core-common/src/main/java/org/apache/kylin/common/PropertiesDelegate.java [hunk 1]
```diff
@@ -39,34 +51,84 @@ public class PropertiesDelegate extends Properties {
...
     public PropertiesDelegate(Properties properties, IExternalConfigLoader configLoader) {
-        if(configLoader != null) this.properties.putAll(configLoader.getProperties());
         this.properties.putAll(properties);
         this.configLoader = configLoader;
+        if (configLoader == null) {
+            this.delegation = this.properties;
+        } else if (configLoader instanceof KylinExternalConfigLoader) {
+            this.delegation = new CompositeMapView(((ICachedExternalConfigLoader)this.configLoader).getPropertyEntries(), this.properties);
+        } else if (configLoader instanceof NacosExternalConfigLoader) {
+            this.delegation = new CompositeMapView(((ICachedExternalConfigLoader)this.configLoader).getPropertyEntries(), this.properties);
```
Comment (zznlime) [inline src/core-common/src/main/java/org/apache/kylin/common/PropertiesDelegate.java:64]: Can this configloader cast into ICachedExternalConfigLoader if it is instanceof NacosExternalConfigLoader?

Comment (hit-lacus) [reply src/core-common/src/main/java/org/apache/kylin/common/PropertiesDelegate.java:64]: Sorry, that is my mistake, I will make it right.

**Comment by hit-lacus** [comment]
image

**Comment by hit-lacus** [comment]
### Patch Applied

```
Starting standalone test @ Thu Sep 29 17:21:46 CST 2022 (1664443306751)
Waiting for possible Shutdown/StopTestNow/HeapDump/ThreadDump message on port 4445
summary +     73 in 00:00:13 =    5.6/s Avg:  3885 Min:  1955 Max:  7941 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary +    400 in 00:00:30 =   13.4/s Avg:  3349 Min:  1841 Max: 24164 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =    473 in 00:00:43 =   11.0/s Avg:  3432 Min:  1841 Max: 24164 Err:     0 (0.00%)

summary +    451 in 00:00:30 =   15.0/s Avg:  2634 Min:  1776 Max:  3431 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =    924 in 00:01:13 =   12.7/s Avg:  3042 Min:  1776 Max: 24164 Err:     0 (0.00%)
summary +    460 in 00:00:30 =   15.2/s Avg:  2611 Min:  1428 Max:  3834 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   1384 in 00:01:43 =   13.4/s Avg:  2899 Min:  1428 Max: 24164 Err:     0 (0.00%)
summary +    485 in 00:00:30 =   16.3/s Avg:  2501 Min:  1134 Max:  3439 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   1869 in 00:02:13 =   14.1/s Avg:  2796 Min:  1134 Max: 24164 Err:     0 (0.00%)
summary +    513 in 00:00:30 =   17.1/s Avg:  2363 Min:  1660 Max:  3453 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   2382 in 00:02:43 =   14.6/s Avg:  2702 Min:  1134 Max: 24164 Err:     0 (0.00%)
summary +    506 in 00:00:30 =   16.9/s Avg:  2356 Min:  1585 Max:  3521 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   2888 in 00:03:13 =   15.0/s Avg:  2642 Min:  1134 Max: 24164 Err:     0 (0.00%)
summary +    510 in 00:00:30 =   16.9/s Avg:  2359 Min:  1652 Max:  3243 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   3398 in 00:03:43 =   15.2/s Avg:  2599 Min:  1134 Max: 24164 Err:     0 (0.00%)
summary +    516 in 00:00:30 =   17.3/s Avg:  2320 Min:  1586 Max:  3453 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   3914 in 00:04:13 =   15.5/s Avg:  2562 Min:  1134 Max: 24164 Err:     0 (0.00%)
summary +    516 in 00:00:30 =   17.2/s Avg:  2306 Min:  1612 Max:  3016 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   4430 in 00:04:43 =   15.7/s Avg:  2533 Min:  1134 Max: 24164 Err:     0 (0.00%)
summary +    337 in 00:00:19 =   18.0/s Avg:  2285 Min:  1383 Max:  3100 Err:     0 (0.00%) Active: 0 Started: 40 Finished: 40
summary =   4767 in 00:05:02 =   15.8/s Avg:  2515 Min:  1134 Max: 24164 Err:     0 (0.00%)
Tidying up ...    @ Thu Sep 29 17:26:48 CST 2022 (1664443608728)
... end of run


Starting standalone test @ Thu Sep 29 17:28:03 CST 2022 (1664443683096)
Waiting for possible Shutdown/StopTestNow/HeapDump/ThreadDump message on port 4445
summary +    403 in 00:00:26 =   15.3/s Avg:  2402 Min:  1679 Max:  4502 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary +    533 in 00:00:30 =   17.8/s Avg:  2249 Min:  1567 Max:  3142 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =    936 in 00:00:56 =   16.6/s Avg:  2315 Min:  1567 Max:  4502 Err:     0 (0.00%)
summary +    510 in 00:00:30 =   17.0/s Avg:  2340 Min:  1629 Max:  3506 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   1446 in 00:01:26 =   16.7/s Avg:  2324 Min:  1567 Max:  4502 Err:     0 (0.00%)
summary +    513 in 00:00:30 =   17.1/s Avg:  2324 Min:  1509 Max:  3338 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   1959 in 00:01:56 =   16.8/s Avg:  2324 Min:  1509 Max:  4502 Err:     0 (0.00%)
summary +    486 in 00:00:30 =   16.2/s Avg:  2512 Min:  1630 Max:  4260 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   2445 in 00:02:26 =   16.7/s Avg:  2361 Min:  1509 Max:  4502 Err:     0 (0.00%)
summary +    559 in 00:00:30 =   18.6/s Avg:  2133 Min:  1542 Max:  2870 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   3004 in 00:02:56 =   17.0/s Avg:  2319 Min:  1509 Max:  4502 Err:     0 (0.00%)
summary +    561 in 00:00:30 =   18.7/s Avg:  2149 Min:  1477 Max:  3624 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   3565 in 00:03:27 =   17.3/s Avg:  2292 Min:  1477 Max:  4502 Err:     0 (0.00%)
summary +    568 in 00:00:30 =   19.0/s Avg:  2103 Min:  1282 Max:  3029 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   4133 in 00:03:56 =   17.5/s Avg:  2266 Min:  1282 Max:  4502 Err:     0 (0.00%)
summary +    556 in 00:00:30 =   18.5/s Avg:  2157 Min:  1307 Max:  3050 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   4689 in 00:04:26 =   17.6/s Avg:  2253 Min:  1282 Max:  4502 Err:     0 (0.00%)
summary +    554 in 00:00:30 =   18.5/s Avg:  2138 Min:  1564 Max:  2839 Err:     0 (0.00%) Active: 40 Started: 40 Finished: 0
summary =   5243 in 00:04:56 =   17.7/s Avg:  2241 Min:  1282 Max:  4502 Err:     0 (0.00%)
summary +    110 in 00:00:05 =   20.8/s Avg:  2204 Min:  1577 Max:  2931 Err:     0 (0.00%) Active: 0 Started: 40 Finished: 40
summary =   5353 in 00:05:02 =   17.7/s Avg:  2240 Min:  1282 Max:  4502 Err:     0 (0.00%)
Tidying up ...    @ Thu Sep 29 17:33:05 CST 2022 (1664443985300)
... end of run
```

**Comment by hit-lacus** [comment]
### After patch applied

image

**Comment by hit-lacus** [comment]
Performance Improve : (31442 - 30028)/30028 = 0.047  = 4.7%

**Comment by hit-lacus** [comment]
- After 
```
summary =  17589 in 00:10:00 =   29.3/s Avg:   170 Min:   119 Max:  4837 Err:     0 (0.00%)
```

- Before
```
summary =  17165 in 00:10:00 =   28.6/s Avg:   174 Min:   125 Max:  5736 Err:     0 (0.00%)
```

**Review by zznlime — COMMENTED** [review]
LGTM

**Pull request merged**